### ISSUE TITLE
Improve `evaluate` for `Interval`s

### ISSUE THREAD

**Issue by lbenet** [JuliaDiff/TaylorSeries.jl#193]
The naive way in which `evaluate` is implemented for `Interval`s may be a limitation, e.g., in relation to https://github.com/JuliaIntervals/TaylorModels.jl/pull/18. For reference, see Sect 2.3 [here](https://easychair.org/publications/open/9Tz3). I think we should add an alternative method, e.g., normalizing the `Interval`s, and take the intersection.

**Comment by mforets** [comment]
Alright, so IIUC you'd like to write an alternative to [this method](https://github.com/JuliaDiff/TaylorSeries.jl/blob/master/src/evaluate.jl#L17) (for the univariate case, say), among one of the 3 methods proposed in that paper. For our case `T::Interval{Float64}` so `suma = suma*dx + a[k]` uses interval arithmetic, and this corresponds to the method "interval arithmetic - no normalized". 

Let me propose what i think is a fourth one - using the bounding property of [Bernstein polynomials](https://en.wikipedia.org/wiki/Bernstein_polynomial), see e.g. [here](https://ac.els-cdn.com/0377042793E0270V/1-s2.0-0377042793E0270V-main.pdf?_tid=a1c3b342-c7ad-4cb0-b97e-195bb9fd368f&acdnat=1549254851_18887d3bbe5e0357249a329946fa8ef1), or [Smith's thesis](https://d-nb.info/1028327854/34). The code in [JuliaReach/BernsteinExpansions.jl](https://github.com/JuliaReach/BernsteinExpansions.jl) can be adapted for this purpose (that we wrote originally with @roccaa). I ignore how does it compare to the other methods, though.

I just experimented with Example 1:

```julia
julia> using TaylorSeries, IntervalArithmetic

julia> x = Taylor1(Interval{Float64}, 6)
 [1, 1] t + 𝒪(t⁷)

julia> R = Interval(0, 6.)
[0, 6]

julia> evaluate(0.1*x^3 - 0.5*x^2 + 1, R) 
[-17, 4.60001]   # this is [-17, 22.6] in Example 1 ... why?? answer below -->

julia> 0.1 * R^3  - 0.5 * R^2 + 1  # Horner's rule is giving better bounds than naive substitution
[-17, 22.6001]
```

I tried the normalisation too; note that:

```python
sage: f = 0.1*x^3 - 0.5*x^2 + 1

sage: f.subs(x= (R[1] - R[0]) * (x+1) / 2 + R[0]).expand()
2.70000000000000*x^3 + 3.60000000000000*x^2 - 0.899999999999999*x - 0.800000000000000 # coincides with the given normalized function
```

But:

```julia
julia> xnew = 2*(x-R.lo) / (R.hi - R.lo) - 1 
 [-1, -1] + [0.333333, 0.333334] t + 𝒪(t⁷)

julia> evaluate(0.1*xnew^3 - 0.5*xnew^2 + 1, Interval(-1, 1.))
[-0.125926, 0.925926] # wrong -- why?

julia> xnew = (R.hi - R.lo) * (x + 1) / 2 + R.lo
 [3, 3] + [3, 3] t + 𝒪(t⁷)

julia> evaluate(0.1*xnew^3 - 0.5*xnew^2 + 1, Interval(-1, 1.))
[-8.00001, 6.40001] # correct? why is the lower bound -8 instead of -4.4 ?
```

**Comment by mforets** [comment]
> normalizing the Intervals, and take the intersection.

Do you have an example where the evaluation without normalisation is better than with it? That being the case, there are 4 cases: `(no normalise, no horner)`, `(normalise, no horner)`, `(no normalise, horner)` and `(normalise, horner)`. Neither of them is better than the other in general, so one has to intersect them.

**Comment by mforets** [comment]
> I ignore how does it compare to the other methods, though.

Just saw that [Section 3.2.4, A. P. Smith Enclosure Methods for Systems of Polynomial Equations and Inequalities] sort of answers my question:

> ... these [Bernstein-like] bounds are in general tighter than those given by interval arithmetic and many centered forms. However it should be noted that the computational effort of generating all the Bernstein coefficients is often much greater than that of these alternatives.

**Comment by lbenet** [comment]
For `TaylorModel`s, things are handled slightly different, in the sense that there may be a remainder of the function, which has to be calculated.

**Comment by lbenet** [comment]
Actually, I had implemented for `TaylorModel1` some way in which the domain evaluated was sliced according to increasing/decreasing parts of the function in question. It uses `IntervalRootFinding` methods; when it works, it yields really sharp bounds, but it is somewhat unstable. 

I am not sure if that can be generalized to `TaylorModelsN` which is important to get sharper bounds in https://github.com/JuliaIntervals/TaylorModels.jl/pull/18.

**Comment by mforets** [comment]
Great, you nailed down the normalization so nicely 💯 

How about higher dimensions? I tried to sum an interval box with a taylor series but it doesn't work:

```julia
julia> TaylorN(3, 3)
 3 + 𝒪(‖x‖⁴)

julia> IntervalBox(0..0, 0..0)
[0, 0] × [0, 0]

julia> IntervalBox(0..0, 0..0) + TaylorN(3, 3)
ERROR: MethodError: no method matching +(::IntervalBox{2,Float64}, ::TaylorN{Int64})
Closest candidates are:
  +(::Any, ::Any, ::Any, ::Any...) at operators.jl:502
  +(::Missing, ::Number) at missing.jl:97
  +(::Taylor1{TaylorN{T<:Union{Real, Complex}}}, ::TaylorN{S<:Union{Real, Complex}}) where {T<:Union{Real, Complex}, S<:Union{Real, Complex}} at /Users/forets/.julia/dev/TaylorSeries/src/arithmetic.jl:183
  ...
Stacktrace:
 [1] top-level scope at none:0
```

**Comment by mforets** [comment]
> For TaylorModels, things are handled slightly different, in the sense that there may be a remainder of the function, which has to be calculated.

In [AGK2018](https://easychair.org/publications/open/9Tz3) they separately calculate `B(P)` and then sum it to `[I]`, the bounding box of the remainder that is  `TaylorModel1.rem`, which is already an interval, so `B(P) + [I]` is computed using interval arithmetic.

Or did i get something wrong? With this reasoning `TaylorModelN.rem` should be an interval box but it is an `Interval{S}` 😕

**Comment by lbenet** [comment]
> How about higher dimensions? I tried to sum an interval box with a taylor series but it doesn't work

The subtlety here is that `TaylorN` is a representation of a multivariate polynomial, which should result in a `Number`; think about the result of `x^2+y^2`. `IntervalBox`es is a vector representation of `Interval`s, so one interval in each independent direction.

What you can do is `evaluate` with an `IntervalBox`:
```julia
julia> r = 3 + TaylorN(1, order=3) * TaylorN(2, order=3)
 3.0 + 1.0 x₁ x₂ + 𝒪(‖x‖⁴)

julia> A = IntervalBox(1..2, 2..4)
[1, 2] × [2, 4]

julia> r(A)
[5, 11]
```

**Comment by mforets** [comment]
> Actually, I had implemented for TaylorModel1 some way in which the domain evaluated was sliced according to increasing/decreasing parts of the function in question. It uses IntervalRootFinding methods; when it works, it yields really sharp bounds, but it is somewhat unstable.

This sounds somehow related to [AGK2018](https://easychair.org/publications/open/9Tz3)'s branch and bound method, no? Algorithm 1 there seems quite complete, we could try to implement it.

**Comment by lbenet** [comment]
(Sorry, I had to leave before.) 

> How about higher dimensions?

What is implemented now in `TaylorSeries` to `evaluate` a `TaylorN` is some sort of Horner's rule. If I recall correctly, each `HomogeneousPolynomial` (monomials in many variables of given degree) is evaluated in some order, which may be the way the coefficients are stored, and the sum over all different monomial degrees is computed. I'll try to dig the code to remember what we did.

> This sounds somehow related to AGK2018's branch and bound method, no? Algorithm 1 there seems quite detailed, we could try to implement it.

Yes, it does sounds as the branch and bound method. The difference is that I was exploiting the intervals where the function is monotonic, where the interval arithmetic evaluation was supposed to be good, and then consider the interval hull. As I said, I in my implementation there were some sort of instabilities in `IntervalRootFinding`. I think that code is still in `TaylorModels`, but maybe commented. In the thesis of Makino there is a nice example on that.

**Comment by dpsanders** [comment]
Note that there is a subtlety:
```
julia> f(t) = -0.8 + -0.9t + 3.6t^2 + 2.7t^3
f (generic function with 1 method)

julia> f(-1..1)
[-4.40001, 6.40001]
```
I.e. the most naive evaluation does even better than Horner. Horner for intervals should be rewritten to take advantage of the fact that even powers of intervals have smaller range.

**Comment by mforets** [comment]
That's a nice example. Using the branch `mforets/normalize_evaluate_Taylor1`:

````julia
julia> using TaylorSeries, IntervalArithmetic

julia> f(t) = -0.8 + -0.9t + 3.6t^2 + 2.7t^3
f (generic function with 1 method)

julia> t = (0..0) + Taylor1(3)
 [1, 1] t + 𝒪(t⁴)

julia> f(t)
 [-0.800001, -0.8] + [-0.900001, -0.9] t + [3.6, 3.60001] t² + [2.7, 2.70001] t³ + 𝒪(t⁴)

julia> evaluate(f(t), (-1..1)) # uses current master
[-8.00001, 6.40001]

julia> evaluate(f(t), (-1..1), horner=false)
[-4.40001, 6.40001]

julia> evaluate(f(t), (-1..1), intersect=true) # normalizes and combines 
[4.59999, 4.60001]
```
The combined case provides an even better bound. Perhaps it should be the new default.

**Comment by dpsanders** [comment]
Can you normalize to [0, 1] too? That seems to give me a better upper bound.

**Comment by mforets** [comment]
Are you sure? Using normalization on `[0, 1]` gives me a worse upper bound, using https://github.com/JuliaDiff/TaylorSeries.jl/pull/195:

```julia
julia> evaluate(f(t), (-1..1), horner=false, normalization="symmetric") # on [-1, 1]
[-4.40001, 6.40001]

julia> evaluate(f(t), (-1..1), horner=false, normalization="asymmetric") # on [0, 1]
[-17.0001, 22.6001]
```

**Comment by dpsanders** [comment]
I think you should not be `evaluate`ing on `(-1..1)` in the asymmetric case?

**Comment by mforets** [comment]
`(-1..1)` is the given interval over which bounds of `f(t)` are searched, so i set it the same for both; normalization is computed internally in the [_normalize function](https://github.com/JuliaDiff/TaylorSeries.jl/pull/195/files#diff-51c18782f21e6732a1cd8fce98d1df05R65).

**Comment by dpsanders** [comment]
The interval to find bounds on of the original function was `R = 0..6`. 
I think the normalized function is evaluated over the corresponding bounds, in this case `0..1` for the asymmetric normalization. But now I'm pretty confused.

**Comment by mforets** [comment]
Hehe OK. I hadn't realized that the polynomial you proposed [here](https://github.com/JuliaDiff/TaylorSeries.jl/issues/193#issuecomment-461869943) was precisely the polynomial obtained *after* the affine map (after normalization on `(-1..1)`).

To use normalization on `(0..1)` i would start over with the original polynomial on `(0..6)`, bring it to `(0..1)` (obtaining a new polynomial) and make the evaluation using horner (or not).

To try out this with `mforets/normalize_evaluate_Taylor1`:

```julia
julia> using TaylorSeries, IntervalArithmetic

julia> t = (0..0) + Taylor1(3)
 [1, 1] t + 𝒪(t⁴)

julia> f(x) = 0.1 * x^3 - 0.5*x^2+1
f (generic function with 1 method)

julia> g = f(t)
 [1, 1] + [-0.5, -0.5] t² + [0.1, 0.100001] t³ + 𝒪(t⁴)

julia> R = 0..6
[0, 6]

julia> evaluate(g, R)
[-17, 4.60001]

julia> evaluate(g, R, normalization="none")
[-17, 4.60001]

julia> evaluate(g, R, normalization="symmetric") # [-1,1]
[-8.00001, 6.40001]

julia> evaluate(g, R, normalization="asymmetric") # [0, 1]
[-17, 4.60001]

julia> evaluate(g, R, intersection=true)
[-4.40001, 4.60001]
```

If this is correct, then the upper bound using normalization on `[0, 1]` is indeed better than using normalization on `[-1,1]` as you claimed before :)

**Comment by dpsanders** [comment]
Yes exactly, thanks!

**Comment by dpsanders** [comment]
It feels to me like the normalization should somehow be dealt with via dispatch, instead of those several `if`s, which feels rather non-Julian.

**Comment by mforets** [comment]
(Please feel free to commit on top of https://github.com/JuliaDiff/TaylorSeries.jl/pull/195/files by dispatching on values instead of kwargs, or to change it in another way, if you prefer.)

**Comment by lbenet** [comment]
> It feels to me like the normalization should somehow be dealt with via dispatch, instead of those several `if`s, which feels rather non-Julian.

A possibility is to suppress those `if`s here and, in `TaylorModel.jl` call the specialized versions of `_evaluate` explicitly, at least for the time being, which do work via dispatch.

**Comment by lbenet** [comment]
Solved by #195. Closing
